Name: Flashlight in Device Menu
Version: 2.0.1-1
Author: Jason Robitaille
Description: Adds a flashlight toggle to the device menu. Requires SysToolsMgr Service.  (MIT license)

--- .orig/usr/lib/luna/system/luna-systemui/app/controllers/devicemenu-assistant.js
+++ /usr/lib/luna/system/luna-systemui/app/controllers/devicemenu-assistant.js
@@ -165,2 +165,3 @@
 		this.controller.get('ime_OnOff').addEventListener(Mojo.Event.tap, this.toggleIMEOnOff.bind(this));
+		this.controller.get('dm_flash').addEventListener(Mojo.Event.tap, this.toggleFlash.bindAsEventListener(this));
 		this.controller.get('imelist').addEventListener(Mojo.Event.listTap,this.handleIMETap.bindAsEventListener(this));
@@ -438,6 +439,31 @@
 		}
 	},
 	
+	toggleFlash: function() {
+		if(this.flashVal==0) {
+			this.flashVal = 100;
+		} else {
+			this.flashVal = 0;
+		}
+		this.controller.serviceRequest('palm://ca.canucksoftware.systoolsmgr', {
+			method: 'flashOn',
+			parameters: {
+				value: this.flashVal
+			},
+			onSuccess: function() {
+				this.setFlashLabel();
+			}.bind(this)
+		});
+	},
+	
+	setFlashLabel: function() {
+		if(this.flashVal==0) {
+			this.controller.get('flashStatus').innerText = "Off";
+		} else {
+			this.controller.get('flashStatus').innerText = "On";
+		}
+	},
+	
 	/*
 	 * Disconnect All Bluetooth Profiles.
 	 */	
@@ -1394,6 +1420,18 @@
 	
 	toggleDeviceMenu: function() {
 		this.clearAllAnimations();
+		this.flashVal = 0;
+		this.controller.serviceRequest('palm://ca.canucksoftware.systoolsmgr', {
+			method: 'flashState',
+			parameters: {},
+			onSuccess: function(response) {
+				this.flashVal = response.value;
+				this.setFlashLabel();
+			}.bind(this),
+			onFailure: function(err) {
+				this.controller.get('flashStatus').innerText = "Error";
+			}.bind(this)
+		});
 		var self = this;
 		if(this.isVisible) {
 			this.isVisible = false;
--- .orig/usr/lib/luna/system/luna-systemui/app/views/devicemenu/devicemenu-scene.html
+++ /usr/lib/luna/system/luna-systemui/app/views/devicemenu/devicemenu-scene.html
@@ -89,6 +89,15 @@
 					
 				</div>	
 				
+				<div class="palm-section-divider"></div>
+
+				<div class="palm-row" id="dm_flash"><div class="palm-row-wrapper">
+					<div class="title truncating-text">
+						<div class="label right" id="flashStatus">&nbsp;</div>
+						<span x-mojo-loc=''>Flashlight</span>
+					</div>
+				</div></div>
+
 				<div class="palm-section-divider"></div>
 				
 				<div id="dm_airplanemode" class="palm-row last menu-end" x-mojo-tap-highlight='momentary'><div class="palm-row-wrapper">
